File: 1 – FMI Data Structures & Algorithms/04–week/04–12 Reverse linked list–2 (recursive).cpp
// https://www.hackerrank.com/challenges/reverse-a-linked-list/problem

SinglyLinkedListNode *reverse(SinglyLinkedListNode *head) {
    // The tail of the list has been reached which
    // will now be the new head of the reversed list
    if (head == nullptr || head->next == nullptr) {
        return head;
    }

    SinglyLinkedListNode *nextNode = head->next;
    head->next = nullptr;

    SinglyLinkedListNode *newHead = reverse(nextNode);
    // reverse the nodes
    nextNode->next = head;
    
    return newHead;
}

File: 1 – FMI Data Structures & Algorithms/04–week/04–12 Reverse linked list–2 (recursive).java
// https://www.hackerrank.com/challenges/reverse-a-linked-list/problem

public static SinglyLinkedListNode reverse(SinglyLinkedListNode head) {
    if (head == null || head.next == null) {
        return head;
    }

    SinglyLinkedListNode remaining = reverse(head.next);

    head.next.next = head;
    head.next = null;

    return remaining;
}